File: primary-identity/vaultClient.js
/**
 * Vault Client - HTTP Client for Vault Service
 * 
 * Communicates with the Vault Service at http://localhost:5000
 * All methods use POST requests with JSON bodies (internal API design)
 * 
 * Error handling:
 *   - 404 from Vault → { success: false, status: 404, error: '...' }
 *   - Network error  → { success: false, status: 503, error: 'Vault service unavailable' }
 *   - 500 from Vault → { success: false, status: 502, error: 'Vault internal error' }
 * 
 * No automatic retries. 5-second timeout on all requests.
 * Does NOT log credential fields (username, password, etc.)
 */

const VAULT_URL = process.env.VAULT_URL || 'http://localhost:5000';
const TIMEOUT_MS = 5000;

/**
 * Make a request to Vault Service with timeout
 * @param {string} endpoint - e.g. '/internal/vault/read'
 * @param {object} body - Request payload
 * @returns {Promise<{success: boolean, status: number, data?: any, error?: string}>}
 */
async function vaultRequest(endpoint, body) {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), TIMEOUT_MS);
    
    try {
        const response = await fetch(`${VAULT_URL}${endpoint}`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(body),
            signal: controller.signal
        });
        
        clearTimeout(timeoutId);
        
        const data = await response.json();
        
        if (response.ok) {
            return { success: true, status: response.status, data };
        }
        
        // Handle Vault errors
        if (response.status === 404) {
            return { success: false, status: 404, error: data.error || 'Not found' };
        }
        
        if (response.status >= 500) {
            console.error(`[VAULT CLIENT] Vault error ${response.status}:`, data.error);
            return { success: false, status: 502, error: 'Vault internal error' };
        }
        
        // Other errors (400, etc.)
        return { success: false, status: response.status, error: data.error || 'Vault request failed' };
        
    } catch (err) {
        clearTimeout(timeoutId);
        
        if (err.name === 'AbortError') {
            console.error('[VAULT CLIENT] Request timeout');
            return { success: false, status: 503, error: 'Vault service timeout' };
        }
        
        console.error('[VAULT CLIENT] Network error:', err.message);
        return { success: false, status: 503, error: 'Vault service unavailable' };
    }
}

/**
 * Read credentials from Vault
 * @param {string} vaultId - e.g. "vault_1"
 * @param {string} appId - e.g. "app_a"
 * @returns {Promise<{success: boolean, status: number, fields?: object, error?: string}>}
 */
async function read(vaultId, appId) {
    console.log(`[VAULT CLIENT] read(vaultId=${vaultId}, appId=${appId})`);
    const result = await vaultRequest('/internal/vault/read', { vaultId, appId });
    
    if (result.success) {
        return { success: true, status: result.status, fields: result.data.fields };
    }
    
    return result;
}

/**
 * Write credentials to Vault
 * @param {string} vaultId
 * @param {string} appId
 * @param {object} fields - { username, password, ... }
 * @returns {Promise<{success: boolean, status: number, error?: string}>}
 */
async function write(vaultId, appId, fields) {
    console.log(`[VAULT CLIENT] write(vaultId=${vaultId}, appId=${appId}) [fields not logged]`);
    return await vaultRequest('/internal/vault/write', { vaultId, appId, fields });
}

/**
 * Update password only (merges into existing fields)
 * @param {string} vaultId
 * @param {string} appId
 * @param {string} newPassword
 * @returns {Promise<{success: boolean, status: number, error?: string}>}
 */
async function updatePassword(vaultId, appId, newPassword) {
    console.log(`[VAULT CLIENT] updatePassword(vaultId=${vaultId}, appId=${appId}) [password not logged]`);
    return await vaultRequest('/internal/vault/update-password', { vaultId, appId, newPassword });
}

/**
 * Delete single credential
 * @param {string} vaultId
 * @param {string} appId
 * @returns {Promise<{success: boolean, status: number, error?: string}>}
 */
async function deleteCredential(vaultId, appId) {
    console.log(`[VAULT CLIENT] delete(vaultId=${vaultId}, appId=${appId})`);
    return await vaultRequest('/internal/vault/delete', { vaultId, appId });
}

/**
 * Delete all credentials for a vault (cascade on user deletion)
 * @param {string} vaultId
 * @returns {Promise<{success: boolean, status: number, error?: string}>}
 */
async function deleteVault(vaultId) {
    console.log(`[VAULT CLIENT] deleteVault(vaultId=${vaultId})`);
    return await vaultRequest('/internal/vault/delete-vault', { vaultId });
}

/**
 * Health check
 * @returns {Promise<boolean>}
 */
async function healthCheck() {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), TIMEOUT_MS);
    
    try {
        const response = await fetch(`${VAULT_URL}/health`, {
            method: 'GET',
            signal: controller.signal
        });
        
        clearTimeout(timeoutId);
        return response.ok;
    } catch (err) {
        clearTimeout(timeoutId);
        return false;
    }
}

module.exports = {
    read,
    write,
    updatePassword,
    delete: deleteCredential,
    deleteVault,
    healthCheck
};

File: primary-identity-python/vault_client.py
"""
Vault Client - Async HTTP Client for Vault Service

Communicates with the Vault Service at http://localhost:5000
All methods use POST requests with JSON bodies (internal API design)

Error handling:
  - 404 from Vault -> {"success": False, "status": 404, "error": "..."}
  - Network error  -> {"success": False, "status": 503, "error": "Vault service unavailable"}
  - 500 from Vault -> {"success": False, "status": 502, "error": "Vault internal error"}

No automatic retries. 5-second timeout on all requests.
Does NOT log credential fields (username, password, etc.)
"""

import os
import httpx

VAULT_URL = os.environ.get("VAULT_URL", "http://localhost:5000")
TIMEOUT_S = 5.0


async def _vault_request(endpoint: str, body: dict) -> dict:
    """Make a request to Vault Service with timeout."""
    try:
        async with httpx.AsyncClient(timeout=TIMEOUT_S) as client:
            response = await client.post(
                f"{VAULT_URL}{endpoint}",
                json=body,
            )

        data = response.json()

        if response.is_success:
            return {"success": True, "status": response.status_code, "data": data}

        # Handle Vault errors
        if response.status_code == 404:
            return {"success": False, "status": 404, "error": data.get("error", "Not found")}

        if response.status_code >= 500:
            print(f"[VAULT CLIENT] Vault error {response.status_code}: {data.get('error')}")
            return {"success": False, "status": 502, "error": "Vault internal error"}

        # Other errors (400, etc.)
        return {"success": False, "status": response.status_code, "error": data.get("error", "Vault request failed")}

    except httpx.TimeoutException:
        print("[VAULT CLIENT] Request timeout")
        return {"success": False, "status": 503, "error": "Vault service timeout"}
    except Exception as err:
        print(f"[VAULT CLIENT] Network error: {err}")
        return {"success": False, "status": 503, "error": "Vault service unavailable"}


async def read(vault_id: str, app_id: str) -> dict:
    """Read credentials from Vault."""
    print(f"[VAULT CLIENT] read(vaultId={vault_id}, appId={app_id})")
    result = await _vault_request("/internal/vault/read", {"vaultId": vault_id, "appId": app_id})

    if result["success"]:
        return {"success": True, "status": result["status"], "fields": result["data"]["fields"]}

    return result


async def write(vault_id: str, app_id: str, fields: dict) -> dict:
    """Write credentials to Vault."""
    print(f"[VAULT CLIENT] write(vaultId={vault_id}, appId={app_id}) [fields not logged]")
    return await _vault_request("/internal/vault/write", {"vaultId": vault_id, "appId": app_id, "fields": fields})


async def update_password(vault_id: str, app_id: str, new_password: str) -> dict:
    """Update password only (merges into existing fields)."""
    print(f"[VAULT CLIENT] updatePassword(vaultId={vault_id}, appId={app_id}) [password not logged]")
    return await _vault_request("/internal/vault/update-password", {"vaultId": vault_id, "appId": app_id, "newPassword": new_password})


async def delete_credential(vault_id: str, app_id: str) -> dict:
    """Delete single credential."""
    print(f"[VAULT CLIENT] delete(vaultId={vault_id}, appId={app_id})")
    return await _vault_request("/internal/vault/delete", {"vaultId": vault_id, "appId": app_id})


async def delete_vault(vault_id: str) -> dict:
    """Delete all credentials for a vault (cascade on user deletion)."""
    print(f"[VAULT CLIENT] deleteVault(vaultId={vault_id})")
    return await _vault_request("/internal/vault/delete-vault", {"vaultId": vault_id})


async def health_check() -> bool:
    """Health check."""
    try:
        async with httpx.AsyncClient(timeout=TIMEOUT_S) as client:
            response = await client.get(f"{VAULT_URL}/health")
            return response.is_success
    except Exception:
        return False
